### ISSUE TITLE
Array type or not?

### ISSUE THREAD

**Issue by todoooo** [CastXML/CastXML#96]
I ran this class through castxml. Please see the attached output. [cast-example.zip](https://github.com/CastXML/CastXML/files/1444371/cast-example.zip)

```
class start
{
	double GetDistance(float *point);
	double GetDistanceByArray(float point[]);
	double GetDistanceBySizedArray(float point[3]);
	double GetDistanceByReferencedArray(float (&point)[3]);
}
```
The parameter in the last method is identified as a reference to an array type (as expected), but I would expect that the parameters in methods 2 & 3 should also be array types, possibly with the max value absent for method 2. Is this a Clang issue or a castxml issue?

**Comment by bradking** [comment]
The C and C++ standards state that array types in function arguments are actually just pointers.  The array syntax is only for human reference and is immediately lowered by the compiler.

**Comment by todoooo** [comment]
In actual fact clang parses these constructs properly with the associated size information. 
https://stackoverflow.com/questions/38516855/how-to-check-if-a-variable-declaration-in-clang-astvisitor-is-an-array
Even multi-dimensional arrays
http://clang-developers.42468.n3.nabble.com/Multi-dimensional-arrays-td4028727.html

Is CastXML likely to support this, as it does for arrays passed by reference?

**Comment by bradking** [comment]
It does parse array types that way for declarations that actually have an array type.  We honor that in `ASTVisitor::OutputConstantArrayType` and `ASTVisitor::OutputIncompleteArrayType`.  That is why the reference-to-array type prints correctly.  However, function parameters in the example don't have array type.  If they did then the output logic would preserve it.

**Comment by bradking** [comment]
Clang does have some APIs to get "AsWritten" information.  This might be available through one of those but it would take some investigation to find.

**Comment by todoooo** [comment]
I'm not following your logic. What does an array type declaration look like then? 

Are you saying this method contains an array parameter
 `double GetDistanceBySizedArray(const float point[3])` 
but this 
`double GetDistanceBySizedArray(float point[3])` 
does not?

**Comment by bradking** [comment]
While `float point[3]` is an array type, in the specific case that it is used as a function parameter type it is treated by the compiler as if it were written `float* point`.  For example:

```console
$ cat test.cpp                                                                   
void f(float (&point)[3]) {}
void f(float point[3]) {}
void f(float* point) {}
$ g++ -c test.cpp
test.cpp: In function ‘void f(float*)’:
test.cpp:3:6: error: redefinition of ‘void f(float*)’
 void f(float* point) {}
      ^
test.cpp:2:6: note: ‘void f(float*)’ previously defined here
 void f(float point[3]) {}
```

**Comment by todoooo** [comment]
I would expect that for duplicate method definitions, but my example class contains none.

I'm using Mummy with CastXML to wrap VTK, so all that matters is the signatures of methods in the generated C# wrappers. The downstream C++ compiler is irrelevant. It appears to me that clang provides additional information about array parameters as it parses the source code and, since CastXML uses clang to generate its output, I would expect to see that information in the XML.

**Comment by bradking** [comment]
I think Clang is erasing this information before CastXML can see it, because the language standard *defines* that array types in function parameters decay to pointers.  If you can find a way to get this information from Clang's API then I'd be happy to use it.

**Comment by bradking** [comment]
See #110 for use of `getOriginalType()` to recover and report the non-decayed type as a separate `original_type=""` attribute.

**Comment by bradking** [comment]
Fixed by #110.
